### PR TITLE
Zach log in page

### PR DESCRIPTION
**Pull request by ZachDFK**
Pull request with master

### CODE REVIEW

File: app/controllers/comments_controller.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+class CommentsController < ApplicationController
+
+  def create
+    @post = Post.find(params[:post_id])
+    
+    @comment = @post.comments.create(comment_params)
```
Comment (kevincox) [inline app/controllers/comments_controller.rb]: Do you have to save the comment or anything?

File: app/controllers/comments_controller.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    redirect_to post_path(@post)
+  end
+
+  private
+    def comment_params
```
Comment (kevincox) [inline app/controllers/comments_controller.rb]: Nitpick, don't indent the private section.

File: app/controllers/pages_controller.rb [hunk 3]
```diff
@@ -0,0 +1,4 @@
+class PagesController < ApplicationController
+  def home
+  end
```
Comment (kevincox) [inline app/controllers/pages_controller.rb]: Is this required?

File: app/controllers/posts_controller.rb [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+  # GET /posts/1
+  # GET /posts/1.json
+  def show
+  end
+
+  # GET /posts/new
+  def new
+    @post = Post.new
+    @post.user = User.find session[:user_id]
```
Comment (kevincox) [inline app/controllers/posts_controller.rb]: Use `current_user`.

File: app/controllers/posts_controller.rb [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+  # GET /posts/1/edit
+  def edit
+  end
+
+  # POST /posts
+  # POST /posts.json
+  def create
+    @post = Post.new(post_params)
+    @post.user = User.find session[:user_id]
```
Comment (kevincox) [inline app/controllers/posts_controller.rb]: `current_user`

File: app/controllers/sessions_controller.rb [hunk 6]
```diff
@@ -0,0 +1,37 @@
+class SessionsController < ApplicationController
+  def login
+  end
+  def login_attempt
+    @login_attempt = login_attempt_params
```
Comment (kevincox) [inline app/controllers/sessions_controller.rb]: Is this ever used?

File: app/models/user.rb [hunk 7]
```diff
@@ -1,2 +1,35 @@
...
+  EMAIL_REGEX =  %r{[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}\z}i
+
+  validates :username, presence: true, uniqueness: true, length: { in: 3..20 }
+  validates :email, presence: true, uniqueness: true, format: EMAIL_REGEX 
+
+  validates :password, confirmation: true, length: {in: 6..20}  
+  validates :password_confirmation, presence: true
+ 
+  has_secure_password    
+  def self.authenticate_with_username_or_email(username_or_email="",login_password="")
```
Comment (kevincox) [inline app/models/user.rb]: Why do these have default arguments? I can't see a use case for not passing them in. If there is a use case we can add them later.

File: app/models/user.rb [hunk 8]
```diff
@@ -1,2 +1,35 @@
...
+  if EMAIL_REGEX.match(username_or_email)
+    user = User.find_by_email(username_or_email)
+  else
+    user = User.find_by_username(username_or_email)
+  end
+  if user && user.authenticate(login_password)
```
Comment (kevincox) [inline app/models/user.rb]: Style nitpick.

``` rb
if user && user.authenticate(login_password)
   user
end
```

If you want to be extra snazzy use try.

``` rb
if user.try :authenticate, login_password
   user
end
```

I can't wait for the elvis operator to ship.

``` rb
if user.?authenticate login_password
   user
end
```

File: app/views/posts/index.html.erb [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+      <th colspan="3"></th>
+    </tr>
+  </thead>
+
+  <tbody>
+    <% @posts.each do |post| %>
+      <tr>
+        <td><%= post.title %></td>
+        <td><%= post.contents %></td>
+        <td><%= User.find(post.user_id).username %></td>
```
Comment (kevincox) [inline app/views/posts/index.html.erb]: This is really inefficient but we can worry about that later if ever.

File: db/migrate/20160310033614_create_comments.rb [hunk 10]
```diff
@@ -0,0 +1,11 @@
+class CreateComments < ActiveRecord::Migration
+  def change
+    create_table :comments do |t|
+      t.integer :user_id
+      t.text :contents
+      t.references :post, index: true, foreign_key: true
```
Comment (kevincox) [inline db/migrate/20160310033614_create_comments.rb]: Oh, this is nice. We should use this for the other "foreign keys". Also is it possible to but "NOT NULL" on these?

**Comment by GoodOlChuck** [comment]
Im getting an error whenever trying to signup - is this still a latent issue, or is this branch not up to date with Zach's local branch? 

`NoMethodError in UsersController#create
undefined method 'password_digest=' for #<User:0x00000005a2a008> Did you mean? password=`

**Comment by kevincox** [comment]
That doesn't sound transient.

**Comment by ZachDFK** [comment]
I'll take a look at all the issues, an try to fix them. @GoodOlChuck, If I remember there was an error with the sign up okay button still linked to a path we removed from routes. It might be why it is looking for UsersController#create but can't find it in routes.

File: app/controllers/comments_controller.rb (hunk 1 above)
Comment (ZachDFK) [inline app/controllers/comments_controller.rb]: It saves by adding it to @post then @post is being updated, or at least that's what it looks like it is doing.

File: app/views/posts/index.html.erb (hunk 9 above)
Comment (ZachDFK) [inline app/views/posts/index.html.erb]: yeah not super proud of it. We could have a method in post model \* edit :
`def author (user_id)`
`author = User find :user_id`
`author.username`
`end`

and in the html
`<%= @post.author %>`

Comment (kevincox) [inline app/views/posts/index.html.erb]: If the foreign keys are set up right you should be able to do `post.user.username`.

What I meant when I said inefficient is that you are doing one query for every post, there is a way to tell rails to combine those all into one query but we can worry about that later.

File: app/controllers/application_controller.rb [hunk 11]
```diff
@@ -1,9 +1,17 @@
...
-    render text:"hi bud"
- end
+  def author_name (user_id)
+  	unless user_id.blank?
+	  author = User.find user_id
+      author.username
+  	else
+  	  '[unknonw]'
+  	end
+  end
```
Comment (kevincox) [inline app/controllers/application_controller.rb]: Can we :fire: this? If the foreign keys are set up right this shouldn't be necessary at all.

File: app/controllers/comments_controller.rb (hunk 1 above)
Comment (kevincox) [inline app/controllers/comments_controller.rb]: Ah I see.

Comment (kevincox) [inline app/controllers/comments_controller.rb]: Wait, but are you ever saving the post?

File: app/models/user.rb [hunk 12]
```diff
@@ -1,2 +1,34 @@
...
+email:     user's email addres it can be used to log in
+=end
+  has_many :posts
+  has_many :comments
+
+  EMAIL_REGEX =  %r{[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}\z}i
+
+  validates :username, presence: true, uniqueness: true, length: { in: 3..20 }
+  validates :email, presence: true, uniqueness: true, format: EMAIL_REGEX 
+  validates :password, confirmation: true, length: {in: 6..20}
```
Comment (kevincox) [inline app/models/user.rb]: Bikeshed: can we make the password max at least 32? Preferably something like 256?

File: db/migrate/20160310021611_create_posts.rb [hunk 13]
```diff
@@ -0,0 +1,10 @@
+class CreatePosts < ActiveRecord::Migration
+  def change
+    create_table :posts do |t|
+      t.string :title, null: false
+      t.text :contents, null: false
+      t.integer :user_id, null: false
```
Comment (kevincox) [inline db/migrate/20160310021611_create_posts.rb]: This should be a foreign key.

File: db/schema.rb [hunk 14]
```diff
@@ -11,17 +11,35 @@
...
   enable_extension "plpgsql"
 
-  create_table "users", force: :cascade do |t|
-    t.string   "username",   null: false
-    t.string   "password",   null: false
-    t.string   "email",      null: false
+  create_table "comments", force: :cascade do |t|
+    t.integer  "user_id"
+    t.text     "contents"
+    t.integer  "post_id"
```
Comment (kevincox) [inline db/schema.rb]: Foreign key.

File: db/schema.rb [hunk 15]
```diff
@@ -11,17 +11,35 @@
...
 
   # These are extensions that must be enabled in order to support this database
   enable_extension "plpgsql"
 
-  create_table "users", force: :cascade do |t|
-    t.string   "username",   null: false
-    t.string   "password",   null: false
-    t.string   "email",      null: false
+  create_table "comments", force: :cascade do |t|
+    t.integer  "user_id"
```
Comment (kevincox) [inline db/schema.rb]: Foreign key.

File: db/schema.rb [hunk 16]
```diff
@@ -11,17 +11,35 @@
...
+    t.datetime "created_at", null: false
+    t.datetime "updated_at", null: false
+  end
+
+  add_index "comments", ["post_id"], name: "index_comments_on_post_id", using: :btree
+
+  create_table "posts", force: :cascade do |t|
+    t.string   "title",      null: false
+    t.text     "contents",   null: false
+    t.integer  "user_id",    null: false
```
Comment (kevincox) [inline db/schema.rb:32]: Foreign key.

**Comment by ZachDFK** [comment]
I'll leave the comment stuff as a seperate issue for now, can you just make sure, the schema.rb is what we are looking for?

File: app/views/posts/show.html.erb [hunk 17]
```diff
@@ -12,14 +12,14 @@
...
-  <%= author_name @post.user_id %>
+  <%= @post.user.username %>
 </p>
 
 <h2>Comments</h2>
 <% @post.comments.each do |comment| %>
 <p>
   <strong>Author:</strong>
-  <%= author_name comment.user_id%>
+  <%= comment.user.username %>
```
Comment (kevincox) [inline app/views/posts/show.html.erb]: :tada:

File: db/schema.rb [hunk 18]
```diff
@@ -17,14 +17,15 @@
   enable_extension "plpgsql"
 
   create_table "comments", force: :cascade do |t|
-    t.integer  "user_id"
     t.text     "contents"
-    t.integer  "post_id"
+    t.integer  "post_id",    null: false
+    t.integer  "user_id",    null: false
```
Comment (kevincox) [inline db/schema.rb]: Both of these should be foreign keys.

File: db/schema.rb [hunk 19]
```diff
@@ -17,14 +17,15 @@
...
     t.datetime "created_at", null: false
     t.datetime "updated_at", null: false
   end
 
   add_index "comments", ["post_id"], name: "index_comments_on_post_id", using: :btree
+  add_index "comments", ["user_id"], name: "index_comments_on_user_id", using: :btree
```
Comment (kevincox) [inline db/schema.rb]: Why using btree? Does it do that for you?

File: db/schema.rb [hunk 20]
```diff
@@ -11,17 +11,41 @@
...
   enable_extension "plpgsql"
 
-  create_table "users", force: :cascade do |t|
-    t.string   "username",   null: false
-    t.string   "password",   null: false
-    t.string   "email",      null: false
+  create_table "comments", force: :cascade do |t|
+    t.text     "contents"
+    t.integer  "post_id",    null: false
+    t.integer  "user_id",    null: false
```
Comment (kevincox) [inline db/schema.rb:22]: These should both be foreign keys.

File: db/schema.rb (hunk 19 above)
Comment (ZachDFK) [inline db/schema.rb]: yep, it was generated by rake using the `references` command

File: db/schema.rb (hunk 20 above)
Comment (ZachDFK) [inline db/schema.rb:22]: yeah that was generated during the rake migrate not exactly sure why

File: db/schema.rb (hunk 19 above)
Comment (kevincox) [inline db/schema.rb]: :+1:

**Comment by ZachDFK** [comment]
I'll fix the merge issue, then do you think that we should post this and start working off master for our other features?

**Comment by ZachDFK** [comment]
Oops, alright I made a mistake with the rebase

File: Gemfile [hunk 21]
```diff
@@ -11,7 +11,8 @@ gem 'rails', '~>4.2.5'
 gem 'sass-rails', '~>5.0'
 gem 'therubyracer', platforms: :ruby
 gem 'turbolinks'
-gem 'uglifier', '~>2.7'
+gem 'uglifier', '>= 1.3.0'
```
Comment (kevincox) [inline Gemfile]: This is reverting my change. Please keep 2.7.

File: db/schema.rb (hunk 20 above)
Comment (kevincox) [inline db/schema.rb:22]: Can this be changed or just leave it like this?

**Comment by kevincox** [comment]
Other than the Gemfile and the foreign keys in the migration I think this is good. Everything except for the db is trivial to tidy up later although the code already looks pretty good so I don't know if we will need that.

**Comment by ZachDFK** [comment]
I'll fix the Gemfile. But that thing in the schema is what is generated by the `t.references` command in the migrate files, according to the [documentation](http://guides.rubyonrails.org/association_basics.html) so I'm not sure there's a way around it unless we use something other than `t.references`.

**Comment by kevincox** [comment]
Ok, I'm happy with I then. ⛴

**Comment by kevincox** [comment]
Is this equivalent to what is now in master?

**Comment by ZachDFK** [comment]
yeah I can close it

**Pull request closed without merging**